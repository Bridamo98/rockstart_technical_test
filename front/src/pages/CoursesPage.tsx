import { useState } from "react";
import { CoursesAPI } from "../api/courses";
import CourseForm from "../components/CourseForm";
import CourseList from "../components/CourseList";
import useCourses from "../hooks/useCourses";
import type { CourseCreateDTO } from "../types";

export default function CoursesPage() {
  const { courses, loading, refresh } = useCourses();
  const [showForm, setShowForm] = useState(false);

  const addCourse = (values: {
    title?: string;
    course_desc?: string;
    instructor_id?: number;
  }) => {
    if (
      !values.title ||
      !values.course_desc ||
      values.instructor_id === undefined
    ) {
      return;
    }
    const data: CourseCreateDTO = {
      title: values.title,
      course_desc: values.course_desc,
      instructor_id: values.instructor_id,
    };
    return CoursesAPI.create(data).then(() => {
      setShowForm(false);
      refresh();
    });
  };

  return (
    <div className="container mx-auto p-6">
      <h1 className="text-3xl font-bold mb-4">Courses</h1>
      <button
        onClick={() => setShowForm(!showForm)}
        className="btn btn-primary mb-4"
      >
        {showForm ? "Cancel" : "New course"}
      </button>

      {showForm && <CourseForm onSubmit={addCourse} />}

      {loading ? <p>Loading â€¦</p> : <CourseList courses={courses} />}
    </div>
  );
}
