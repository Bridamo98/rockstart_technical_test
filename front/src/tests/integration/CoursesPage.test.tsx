import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { expect, test } from "vitest";
import CoursesPage from "../../pages/CoursesPage";
import { resetMockState } from "../mocks/handlers";
import { renderWithProviders } from "../testUtils";

beforeEach(resetMockState);

test("flujo de crear & lista curso (snapshot)", async () => {
  const { asFragment } = renderWithProviders(<CoursesPage />);
  const user = userEvent.setup();

  expect(asFragment()).toMatchSnapshot("CoursesPage-estado-inicial");

  await user.click(screen.getByRole("button", { name: /new course/i }));
  await user.type(screen.getByPlaceholderText(/title/i), "E2E");
  await user.type(screen.getByPlaceholderText(/description/i), "desc");
  await user.click(screen.getByRole("button", { name: /^save$/i }));

  await waitFor(() => expect(screen.getByText("E2E")).toBeInTheDocument());

  expect(asFragment()).toMatchSnapshot("CoursesPage-tras-crear-curso");
});

test("cancelar formulario vuelve al estado anterior", async () => {
  const { asFragment } = renderWithProviders(<CoursesPage />);
  const user = userEvent.setup();

  await user.click(screen.getByRole("button", { name: /new course/i }));
  expect(screen.getByRole("button", { name: /cancel/i })).toBeInTheDocument();

  await user.click(screen.getByRole("button", { name: /cancel/i }));
  expect(asFragment()).toMatchSnapshot("CoursesPage-cancelado");
});
