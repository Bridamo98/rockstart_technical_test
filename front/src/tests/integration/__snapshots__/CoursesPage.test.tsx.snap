// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cancelar formulario vuelve al estado anterior > CoursesPage-cancelado 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6"
  >
    <h1
      class="text-3xl font-bold mb-4"
    >
      Courses
    </h1>
    <button
      class="btn btn-primary mb-4"
    >
      New course
    </button>
    <ul
      class="space-y-2"
    />
  </div>
</DocumentFragment>
`;

exports[`flujo de crear & lista curso (snapshot) > CoursesPage-estado-inicial 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6"
  >
    <h1
      class="text-3xl font-bold mb-4"
    >
      Courses
    </h1>
    <button
      class="btn btn-primary mb-4"
    >
      New course
    </button>
    <p>
      Loading â€¦
    </p>
  </div>
</DocumentFragment>
`;

exports[`flujo de crear & lista curso (snapshot) > CoursesPage-tras-crear-curso 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6"
  >
    <h1
      class="text-3xl font-bold mb-4"
    >
      Courses
    </h1>
    <button
      class="btn btn-primary mb-4"
    >
      New course
    </button>
    <ul
      class="space-y-2"
    >
      <li
        class="p-4 bg-white rounded shadow hover:bg-gray-50"
      >
        <a
          href="/courses/1"
        >
          <h2
            class="font-semibold"
          >
            E2E
          </h2>
          <p
            class="text-sm text-gray-600"
          >
            desc
          </p>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
